app/models/ems_refresh/save_inventory.rb

```diff
@@ -181,6 +181,29 @@ def save_vms_inventory(ems, hashes, target = nil)
     end
   end
 
+  # Convert all mapped hashes into actual tags and associate them with the object.
+  # The hash[:tags] object should be an array of hashes, probably created by
+  # the ContainerLabelTagMapping.map_labels method. Each hash in the array should
+  # have the following basic structure:
+  #
+  # {:category_tag_id=>139, :entry_name=>"foo", :entry_description=>"bar"}
+  #
+  # For now, the object must be a VmOrTemplate object.
+  #
+  def save_tags_inventory(object, hash)
+    return if hash.blank?
+    return unless object.kind_of?(VmOrTemplate)
+
+    tags = hash[:tags].collect do |tag_hash|
+      ContainerLabelTagMapping.find_or_create_tag(tag_hash)
```

Comment:
hm, this method looks super ineffective for being run multiple times for each vm_or_template (huge n+1)

https://github.com/Ladas/manageiq/blob/3231e3993e01b3b8078ec5efada9c9c66b20479c/app/models/container_label_tag_mapping.rb#L75-L75

@kbrock seems to be super ineffective to be call here find_or_create_tag